Code/GraphMol/FileParsers/GeneralFileReader.h

```diff
@@ -0,0 +1,213 @@
+#include <string>
+#include <vector>
+
+#include "MolSupplier.h"
+
+namespace RDKit {
+namespace GeneralMolSupplier {
+struct SupplierOptions {
+  bool takeOwnership = true;
+  bool sanitize = true;
+  bool removeHs = true;
+  bool strictParsing = true;
+
+  std::string delimiter = "\t";
+  int smilesColumn = 0;
+  int nameColumn = 1;
+  bool titleLine = true;
+
+  std::string nameRecord = "";
+  int confId2D = -1;
+  int confId3D = 0;
+};
+//! current supported file formats
+std::vector<std::string> supportedFileFormats{"sdf", "mae", "maegz", "smi",
+                                              "csv", "txt", "tsv",   "tdt"};
+//! current supported compression formats
+std::vector<std::string> supportedCompressionFormats{"gz"};
+
+//! determines whether file and compression format are valid
+bool validate(std::string& fileFormat, std::string& compressionFormat) {
+  bool flag_fileFormat = true;
+  bool flag_compressionFormat = true;
+
+  //! Case 1: No compression format
+  if (compressionFormat.empty()) {
+    //! Unconventional format types
+    if (fileFormat.compare("maegz") == 0) {
+      fileFormat = "mae";
+      compressionFormat = "gz";
+      return true;
+    }
+
+    //! Usual file formats
+    return std::find(supportedFileFormats.begin(), supportedFileFormats.end(),
+                     fileFormat) != supportedFileFormats.end();
+  }
+  //! Case 2: With compression format
+  else {
+    flag_fileFormat =
+        std::find(supportedFileFormats.begin(), supportedFileFormats.end(),
+                  fileFormat) != supportedFileFormats.end();
+    flag_compressionFormat =
+        std::find(supportedCompressionFormats.begin(),
+                  supportedCompressionFormats.end(),
+                  compressionFormat) != supportedCompressionFormats.end();
+
+    //! Case A: Valid file format and valid compression format
+    if (flag_fileFormat && flag_compressionFormat) {
+      return true;
+    }
+
+    //! Case B, C: Valid file format but not valid compression format
+    //! 						or invalid file format
+    //! and valid compression format
+    else if ((flag_fileFormat && !flag_compressionFormat) ||
+             (!flag_fileFormat && flag_compressionFormat)) {
+      return false;
+    }
+
+    //! Case D: Invalid file format and invalid compression format
+    else {
+      //! it is possible that we have read a file with name *.2.txt (for
+      //! example) thus fileformat = "2" and compressionformat = "txt", so we
+      //! try to set the file format as the compression format and the
+      //! compression format as an empty string. Then we check of validity.
+      fileFormat = compressionFormat;
+      compressionFormat = "";
+      return std::find(supportedFileFormats.begin(), supportedFileFormats.end(),
+                       fileFormat) != supportedFileFormats.end();
+    }
+  }
+}
+
+//! returns the file name givens the file path
+std::string getFileName(const std::string path) {
+  char delimiter = '/';
```

Comment:
This is Unix-specific, yes? Windows uses \ as the path delimiter. Is there a Boost function for this?